search/ranking_utils.hpp

```diff
@@ -120,85 +115,98 @@ enum NameScore
   NAME_SCORE_COUNT
 };
 
+struct NameScores
+{
+  NameScores() = default;
+  NameScores(NameScore nameScore, ErrorsMade const & errorsMade)
+    : m_nameScore(nameScore), m_errorsMade(errorsMade)
+  {
+  }
+
+  static NameScores BestScores(NameScores const & lhs, NameScores const & rhs);
+
+  bool operator==(NameScores const & rhs)
+  {
+    return m_nameScore == rhs.m_nameScore && m_errorsMade == rhs.m_errorsMade;
+  }
+
+  NameScore m_nameScore = NAME_SCORE_ZERO;
+  ErrorsMade m_errorsMade;
+};
+
 // Returns true when |s| is a stop-word and may be removed from a query.
 bool IsStopWord(strings::UniString const & s);
 
 // Normalizes, simplifies and splits string, removes stop-words.
 void PrepareStringForMatching(std::string const & name, std::vector<strings::UniString> & tokens);
 
 template <typename Slice>
-NameScore GetNameScore(std::string const & name, Slice const & slice)
-{
-  if (slice.Empty())
-    return NAME_SCORE_ZERO;
-
-  std::vector<strings::UniString> tokens;
-  SplitUniString(NormalizeAndSimplifyString(name), base::MakeBackInsertFunctor(tokens), Delimiters());
-  return GetNameScore(tokens, slice);
-}
-
-template <typename Slice>
-NameScore GetNameScore(std::vector<strings::UniString> const & tokens, Slice const & slice)
+NameScores GetNameScores(std::vector<strings::UniString> const & tokens, Slice const & slice)
 {
   if (slice.Empty())
-    return NAME_SCORE_ZERO;
+    return {};
 
   size_t const n = tokens.size();
   size_t const m = slice.Size();
 
   bool const lastTokenIsPrefix = slice.IsPrefix(m - 1);
 
-  NameScore score = NAME_SCORE_ZERO;
+  NameScores scores;
   for (size_t offset = 0; offset + m <= n; ++offset)
   {
+    ErrorsMade totalErrorsMade;
     bool match = true;
     for (size_t i = 0; i < m - 1 && match; ++i)
-      match = match && impl::FullMatch(slice.Get(i), tokens[offset + i]);
+    {
+      auto errorsMade = impl::GetErrorsMade(slice.Get(i), tokens[offset + i]);
+      match = match && errorsMade.IsValid();
+      totalErrorsMade += errorsMade;
+    }
+
     if (!match)
       continue;
 
-    bool const fullMatch = impl::FullMatch(slice.Get(m - 1), tokens[offset + m - 1]);
-    bool const prefixMatch =
-        lastTokenIsPrefix && impl::PrefixMatch(slice.Get(m - 1), tokens[offset + m - 1]);
-    if (!fullMatch && !prefixMatch)
+    auto const prefixErrorsMade =
```

Comment:
nit: don't call GetPrefixErrorsMade if lastTokenIsPrefix is false

Reply:
done